server/src/main/java/org/opensearch/common/concurrent/GatedAutoCloseable.java

```diff
@@ -0,0 +1,43 @@
+/*
```

Comment:
Discussed in more detail offline. We wouldn't be able to replace the `Gated*` classes directly since they implement additional interfaces, but it's fair to ask if `AbstractRefCounted` can replace `OneWayGate`. 

That is certainly doable but it would involve creating a subclass of AbstractRefCounted that overrides nearly all the functionality in the superclass. At that point, the parent-child relationship becomes very thin so I chose to keep the notion of a "gate" and "ref-count" separate.